core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/definitions/user/BasicAuthenticationRulesSimilarity.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ *    This file is part of ReadonlyREST.
+ *
+ *    ReadonlyREST is free software: you can redistribute it and/or modify
+ *    it under the terms of the GNU General Public License as published by
+ *    the Free Software Foundation, either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    ReadonlyREST is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
+ */
+package tech.beshu.ror.accesscontrol.blocks.definitions.user
+
+import tech.beshu.ror.accesscontrol.blocks.rules.auth.AuthKeyHashingRule.HashedCredentials
+import tech.beshu.ror.accesscontrol.blocks.rules.auth.AuthKeyHashingRule.HashedCredentials.{HashedOnlyPassword, HashedUserAndPassword}
+import tech.beshu.ror.accesscontrol.blocks.rules.auth.base.BasicAuthenticationRule
+import tech.beshu.ror.accesscontrol.blocks.rules.auth.*
+import tech.beshu.ror.utils.Similarity
+import tech.beshu.ror.utils.Similarity.SimilarityMeasure
+
+private[user] object BasicAuthenticationRulesSimilarity {
+
+  private type SupportedRule = AuthKeyRule | AuthKeyUnixRule | AuthKeyHashingRule
+
+  def similarity(x: BasicAuthenticationRule[?], y: BasicAuthenticationRule[?]): SimilarityMeasure = {
```

Comment:
This implementation looks very nice, but I'm not sure if this is what the user needs. 
IMO we should compare the `username` fields. In code, this is represented by `UserIdPatterns` case class. 

We already have the `withUserIdParamsCheck`. It checks if the user defined in the auth rule matches the pattern defined in the `usernme` field. The user noticed that we have no validation of `usenname` patterns uniqueness. 
In the client's example the validation won't help, because username in the auth rule is hashed. 
But it seems there should be another validation, which can be defined like this:

- if there are at least two the same usename pattens (without wildcard) it should failed to validate

And that's it. Is it? Or I'm missing sth?

TBH, the "username" seems to be a redundant setting. In case of auth rules with local users it seems that it is not necessary. In case of rules like LDAP auth we can use it to cut out some users. In case of local users it makes no sense.